File: sdks/java/basys.sdk/backends/org/eclipse/basyx/aas/backend/connected/property/ConnectedMapProperty.java
package org.eclipse.basyx.aas.backend.connected.property;

import java.util.Collection;
import java.util.Map;

import org.eclipse.basyx.aas.api.exception.ServerException;
import org.eclipse.basyx.aas.api.exception.TypeMismatchException;
import org.eclipse.basyx.aas.api.resources.IMapProperty;
import org.eclipse.basyx.aas.api.resources.PropertyType;
import org.eclipse.basyx.vab.core.proxy.VABElementProxy;

/**
 * Connects to a PropertySingleValued as specified by DAAS containing a map
 * 
 * @author schnicke
 *
 */
public class ConnectedMapProperty extends ConnectedProperty implements IMapProperty {

	public ConnectedMapProperty(String path, VABElementProxy proxy) {
		super(PropertyType.Map, path, proxy);
	}

	@Override
	public Object getValue(String key) throws TypeMismatchException, ServerException {
		return getMap().get(key);
	}

	@Override
	public void put(String key, Object value) throws ServerException {
		try {
			getProxy().updateElementValue(constructPath("value/" + key), value);
		} catch (Exception e) {
			e.printStackTrace();
		}
	}

	@Override
	public void set(Map<String, Object> map) throws ServerException {
		try {
			getProxy().updateElementValue(constructPath("value"), map);
		} catch (Exception e) {
			e.printStackTrace();
		}
	}

	@Override
	public Collection<String> getKeys() throws TypeMismatchException, ServerException {
		return getMap().keySet();
	}

	@Override
	public Integer getEntryCount() throws TypeMismatchException, ServerException {
		return getMap().entrySet().size();
	}

	@Override
	public void remove(String key) throws ServerException, TypeMismatchException {
		try {
			getProxy().deleteElement(constructPath("value/" + key));
		} catch (Exception e) {
			e.printStackTrace();
		}
	}

	@SuppressWarnings("unchecked")
	private Map<String, Object> getMap() {
		return ((Map<String, Object>) getProxy().readElementValue(constructPath("value")));
	}

	@Override
	public void setValue(Object obj) {
		// TODO Auto-generated method stub
		
	}

	@Override
	public Object getValue() {
		// TODO Auto-generated method stub
		return null;
	}

	@Override
	public void setValueId(Object obj) {
		// TODO Auto-generated method stub
		
	}

	@Override
	public Object getValueId() {
		// TODO Auto-generated method stub
		return null;
	}

	@Override
	public void setSemanticId(Object obj) {
		// TODO Auto-generated method stub
		
	}

	@Override
	public Object getSemanticId() {
		// TODO Auto-generated method stub
		return null;
	}

}

File: sdks/c++/basys.sdk.cc/src/aas/backend/connected/aas/submodelelement/property/ConnectedMapProperty.cpp
/*
 * ConnectedMapProperty.cpp
 *
 *      Author: wendel
 */

#include "ConnectedMapProperty.h"
#include "vab/core/util/VABPath.h"

namespace basyx {
namespace aas {
namespace backend {
namespace connected {

using namespace submodelelement::property;

ConnectedMapProperty::ConnectedMapProperty(std::shared_ptr<vab::core::proxy::IVABElementProxy> proxy) :
  ConnectedProperty(PropertyType::Map, proxy)
{}

basyx::any ConnectedMapProperty::getValue(const std::string & key) const
{
  return this->getMap().at(key);
}

void ConnectedMapProperty::put(const std::string & key, const basyx::any & value) const
{
  basyx::vab::core::VABPath path(PropertyPaths::VALUE);
  path.append(key);
  this->setProxyValue(path, value);
}

void ConnectedMapProperty::set( const basyx::objectMap_t & map ) const
{
  this->setProxyValue(PropertyPaths::VALUE, map);
}

basyx::objectCollection_t ConnectedMapProperty::getKeys() const
{
  auto map = this->getMap();
  basyx::objectCollection_t keys;
  keys.reserve(map.size());

  for ( auto entry : map ) {
    keys.push_back(entry.first);
  }

  return keys;
}

int ConnectedMapProperty::getEntryCount() const 
{
  return this->getMap().size();
}

void ConnectedMapProperty::remove(const std::string & key) const
{
  basyx::vab::core::VABPath path(PropertyPaths::VALUE);
  path.append(key);
  this->getProxy()->deleteElement(path);
}

basyx::objectMap_t ConnectedMapProperty::getMap() const
{
  auto map = this->getProxy()->readElementValue(PropertyPaths::VALUE).Get<basyx::objectMap_t>();
  return map;
}

}
}
}
}
